pylinalg/obj/transform.py

```diff
@@ -0,0 +1,137 @@
+        quaternion (format: x, y, z, w) describing the rotation from child to
+        parent.)
+    scale : ndarray, [3]
+        The scale of the parent expressed in child's units. (A axis-wise scaling
+        to map units from child to parent.)
+
+    """
+
+    # some constants for syntax sugar
+    FORWARD = np.array((0, 0, 1))
+    BACKWARD = np.array((0, 0, -1))
+    UP = np.array((0, 1, 0))
+    DOWN = np.array((0, -1, 0))
+    RIGHT = np.array((1, 0, 0))
+    LEFT = np.array((-1, 0, 0))
+
+    def __init__(self, /, *, position=None, orientation=None, scale=None) -> None:
+        self._position = position
+        self._orientation = orientation
+        self._scale = scale
+
+        if position is None:
+            self._position = np.zeros(3)
+
+        if orientation is None:
+            self._orientation = np.array((0, 0, 0, 1))
+
+        if scale is None:
+            self._scale = np.ones(3)
+
+    def as_matrix(self):
+        """
+        The affine transforms matrix
+
+        The returned matrix is a transformation matrix that converts
+        points in homogeneous coordinates according to the current parameters
+        of this AffineTransform object. Typically, this maps from an object's
+        child frame into an object's parent frame.
+        """
+
+        result = qt.quaternion_to_matrix(self._orientation)
+        result[:, :3] *= self._scale.reshape(1, 3)
+        result[:3, -1] = self._position
+
+        return result
+
+    def inverse(self):
+        """
+        Invert this affine transformation.
+
+
+        Note
+        ----
+        The problem with this implementation is that this transform and its
+        inverse don't remain in sync, i.e., if we store the returned transform
+        somewhere and this transform updates later, the stored transform does no
+        longer invert this transform. -- We could work around this by
+        introducing `Transform.inverse_transform` and a
+        `InverseTransform(Transform)` object if desired.
+        """
+
+        scale = 1 / self.scale
+        orientation = qt.quaternion_inverse(self.orientation)
+        position = -(
+            scale * (qt.quaternion_to_matrix(orientation)[:3, :3] @ self._position)
+        )
+
+        return AffineTransform(position=position, orientation=orientation, scale=scale)
+
+    def __array__(self, dtype=None):
+        return self.as_matrix().astype(dtype)
+
+    def __matmul__(self, other):
+        if not isinstance(other, AffineTransform):
+            raise NotImplementedError()
+
+        rotation = qt.quaternion_to_matrix(self.orientation)[:3, :3]
+        position = self.position + self.scale * (rotation @ other.position)
+
+        orientation = qt.quaternion_multiply(other._orientation, self._orientation)
+
+        scale = other.scale * self.scale
+
+        return AffineTransform(position=position, orientation=orientation, scale=scale)
```

Comment:
The current implementation assumes that the transform can always be described with a separate position, orientation, and scale. It is thus a subset of the affine transform, because it does not include shear. But this assumption breaks down when you compose two transforms, one of which has a rotation and the other a non-isotropic scale, since that will introduce shear.

This means that we can have multiple transforms, and set their properties as we please, but as soon as we compose two transforms, we don't know whether we need a matrix to describe it (instead of the properties).

 One way to solve this could be to produce a MatrixTransform when two transforms are composed. (or perhaps that should be called an AffineTransform, and this transform should have a different name, since it's really a subset of the Affine transform.)

Or, instead of properties it could have methods (`scale`, `rotate`, `translate`) and internally it would use the properties directly, or a matrix when it's needed.

Yet another solution could be to only allow non-isotropic scale at the leaf nodes?

Reply:
A variation of the first solution would be to check the case where shear is introduced, and only return a different `Transformation` type in that case.

Reply:
Yeah, that's true. Not sure how I missed that. How about we add `AffineTransform.shear` and - for completeness - a `AffineTransform.projection` component? This way, we should be able to represent any affine transformation.

We can use a generic implementation of `AffineTransform @ AffineTransform` for now, where we construct two matrices internally, multiply them, and then decompose the result. Most of this is already implemented in the functional API, and all that is left is to update `matrix_decompose` to return a shear component as well. On top, we add a fast path for special cases where we know efficient solutions (e.g. multiplying with isotropic scale).

We would also have to decide in which coordinate system we want to specify each component. Currently, all parameters are specified in the source's frame (i.e., the parent frame), but we might want to choose something else depending on what you find intuitive.